core/provers.go

```diff
@@ -20,36 +23,48 @@ type ProverI interface {
+	QueryClientConsensusStateWithProof(ctx QueryContext, dstClientConsHeight ibcexported.Height) (*clienttypes.QueryConsensusStateResponse, error)
 
-	// QueryHeader returns the header corresponding to the height
-	QueryHeader(height int64) (out HeaderI, err error)
+	// QueryClientStateWithProof returns the ClientState and its proof
+	QueryClientStateWithProof(ctx QueryContext) (*clienttypes.QueryClientStateResponse, error)
 
-	// QueryLatestHeader returns the latest header from the chain
-	QueryLatestHeader() (out HeaderI, err error)
+	// QueryConnectionWithProof returns the Connection and its proof
+	QueryConnectionWithProof(ctx QueryContext) (*conntypes.QueryConnectionResponse, error)
 
-	// GetLatestLightHeight returns the latest height on the light client
-	GetLatestLightHeight() (int64, error)
+	// QueryChannelWithProof returns the Channel and its proof
+	QueryChannelWithProof(ctx QueryContext) (chanRes *chantypes.QueryChannelResponse, err error)
 
+	// QueryPacketCommitmentWithProof returns the packet commitment and its proof
+	QueryPacketCommitmentWithProof(ctx QueryContext, seq uint64) (comRes *chantypes.QueryPacketCommitmentResponse, err error)
+
+	// QueryPacketAcknowledgementCommitmentWithProof returns the packet acknowledgement commitment and its proof
+	QueryPacketAcknowledgementCommitmentWithProof(ctx QueryContext, seq uint64) (ackRes *chantypes.QueryPacketAcknowledgementResponse, err error)
+}
+
+// LightClient provides functions for creating and updating on-chain light clients on the counterparty chain
+type LightClient interface {
 	// CreateMsgCreateClient creates a CreateClientMsg to this chain
-	CreateMsgCreateClient(clientID string, dstHeader HeaderI, signer sdk.AccAddress) (*clienttypes.MsgCreateClient, error)
+	CreateMsgCreateClient(clientID string, dstHeader Header, signer sdk.AccAddress) (*clienttypes.MsgCreateClient, error)
 
-	// SetupHeader creates a new header based on a given header
-	SetupHeader(dst LightClientIBCQueryierI, baseSrcHeader HeaderI) (HeaderI, error)
+	// GetLatestFinalizedHeader returns the latest finalized header on this chain
+	// The returned header is expected to be the latest one of headers that can be verified by the light client
+	GetLatestFinalizedHeader() (latestFinalizedHeader Header, err error)
 
-	// UpdateLightWithHeader updates a header on the light client and returns the header and height corresponding to the chain
-	UpdateLightWithHeader() (header HeaderI, provableHeight int64, queryableHeight int64, err error)
+	// SetupHeadersForUpdate returns the finalized header and any intermediate headers needed to apply it to the client on the counterpaty chain
+	// The order of the returned header slice should be as: [<intermediate headers>..., <update header>]
+	// if the header slice's length == 0 and err == nil, the relayer should skips the update-client
+	SetupHeadersForUpdate(dstChain ChainInfoICS02Querier, latestFinalizedHeader Header) ([]Header, error)
```

Comment:
Can we change `LightClient` interface as below?
- `GetLatestFinalizedHeader` -> `GetLatestFinalizedHeight`
- The second argument of `SetupHeadersForUpdate`: `Header` -> `Height`

If we change codes like this, in the case where we already have a target height, we can directly use it for `SetupHeadersForUpdate` without calling `GetLatestFinalizedHeader`.

Reply:
I understand that your suggestion assumes that the past finalised headers can be retrieved using height at any point in time, it is somewhat difficult to satisfy this in ethereum at least.

In ethereum's sync protocol, it is difficult to get a finalised header corresponding to any height via the standard API: ref. https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/p2p-interface.md#light_client_finality_update

> in the case where we already have a target height

Although `GetLatestFinalisedHeader` must be called, we can avoid unnecessary (and expensive) update-client(s) if `SetupHeadersForUpdate` returns empty.